[iPadOS] Outlook native app: channels tab does not handle trackpad two-finger pans to scroll

whsieh opened the pull request:
#### d259d7a2d4e9bd22d8012ab3c9947dd54a366aca
<pre>
[iPadOS] Outlook native app: channels tab does not handle trackpad two-finger pans to scroll
<a href="https://bugs.webkit.org/show_bug.cgi?id=283931">https://bugs.webkit.org/show_bug.cgi?id=283931</a>
<a href="https://rdar.apple.com/139834901">rdar://139834901</a>

Reviewed by Abrar Rahman Protyasha.

On iPad, when scrolling via 2-finger swipe on a connected trackpad, `wheel` events are dispatched to
the page via the `WebPage::dispatchWheelEventWithoutScrolling` IPC message. As the name suggests,
this only dispatches `wheel` DOM events to the page, and returns a boolean result back to the UI
process indicating whether or not scrolling should be allowed to commence. UIKit will then only
drive the default scroll animation on the native scroll view under the pointer if the event was not
`handled`.

Whether the event is handled or not depends on the return value of `EventHandler::handleWheelEvent`.
In the case of the Outlook native app&apos;s Channels tab, which is comprised of a large full-viewport
subscrollable overflow container, it returns `true` only because the body and root elements have
`overscroll-behavior: none;`, causing us to exit early in `handleWheelEventInternal` with `true`:

```
HandleUserInputEventResult EventHandler::handleWheelEventInternal(…)
{
…
    auto adjustedWheelEvent = event;
    auto filteredDelta = adjustedWheelEvent.delta();
    filteredDelta = view-&gt;deltaForPropagation(filteredDelta);
    if (view-&gt;shouldBlockScrollPropagation(filteredDelta))
        return true;                                            // &lt;---
```

...and subsequently causes us to block scrolling on the subscrollable container. In this same
codepath on macOS, we also return `true` here, but the key difference is that treating the wheel
event as `handled := true` here doesn&apos;t simply block scrolling — that only happens if we&apos;ve blocked
scrolling due to `preventDefault` being called.

To fix this impedance mismatch between the return value of `handleWheelEvent()` and how it&apos;s
interpreted within `WebPage::dispatchWheelEventWithoutScrolling`, we plumb the `EventHandling`
result alongside `HandleUserInputEventResult` up from WebCore to WebKit as a `std::pair`, and
consult the event handling result (i.e. whether or not `EventHandling::DefaultPrevented` is set) in
addition to checking whether the event was handled.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleWheelEvent):

Make this return a `std::pair&lt;HandleUserInputEventResult, OptionSet&lt;EventHandling&gt;&gt;`, and plumb it
through to the client layer (see `WebPage` changes below) so that we can check if the default was
prevented (not just if the event was handled) in `dispatchWheelEventWithoutScrolling`.

(WebCore::EventHandler::passWheelEventToWidget):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::wheelEvent):
* Source/WebCore/page/mac/EventHandlerMac.mm:
(WebCore::EventHandler::wheelEvent):
(WebCore::EventHandler::passWheelEventToWidget):
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::shouldBlockScrollPropagation const):

Drive-by fix: rewrite this method to be more readable:

-   Cache the results of calling `(horizontal|vertical)OverscrollBehaviorPreventsPropagation` in
    local variables instead of calling the respective methods multiple times.

-   Split the return value into several early returns that achieve the same result, but (more
    clearly) represent the scenarios in which this is true: either both axes stop propagation, or
    only one axis stops propagation but the biased delta amount for the opposite axis is 0.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollView:handleScrollUpdate:completion:]):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::dispatchWheelEvent):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::handleWheelEvent):
(WebKit::WebPage::wheelEvent):
(WebKit::WebPage::dispatchWheelEventWithoutScrolling):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Rename this variable to reflect that this result flag represents whether or not scrolling should be
prevented, not just whether the event was handled.

* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
(-[WKWebView synchronouslyHandleScrollEventWithPhase:location:delta:]):
(TEST(WKScrollViewTests, AsynchronousWheelEventHandling)):
(TEST(WKScrollViewTests, OverscrollBehaviorShouldNotPreventScrolling)):

Add a new API test to exercise this change, by verifying that scrolling is not prevented when
dispatching wheel events over an overflow scrollable region inside of a root and body element with
`overscroll-behavior: contain;`. Also verify that scrolling is (correctly) prevented if the page
explicitly calls `preventDefault()` on the wheel event.

Canonical link: <a href="https://commits.webkit.org/287293@main">https://commits.webkit.org/287293@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c5e147fdafcdcf5fa9be33fcb69f575367715fc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79054 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58090 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83694 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30269 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81187 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67199 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6360 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61877 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19794 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82121 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51917 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72063 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42180 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49268 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26164 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28633 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70383 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26582 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85080 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4408 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70114 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6543 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67935 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69363 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17289 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13412 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12196 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6330 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12162 "Build is in progress. Recent messages:OS: Sequoia (15.0.1), Xcode: 16.0; Checked out pull request; Running scan-build") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6291 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9742 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8083 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/73228c49b2df359aa993fcfde0ac94ba800c6845. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37333

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/57835524493ebb8f4fdba16e08a8aaf296d8397c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37333

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ff7bb06ea85a0f002e028a0f0739065d9840a0d8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37333

aprotyas commented on Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
```diff
@@ -172,6 +172,30 @@ inline static UIScrollPhase legacyScrollPhase(WKBEScrollViewScrollUpdatePhase ph
+
+    RetainPtr update = createScrollUpdate(phase, location, delta);
+    [self scrollView:self.scrollView handleScrollUpdate:update.get() completion:^(BOOL prevented) {
+        scrollingPrevented = prevented;
```
Implicit `BOOL` to `bool`, I suppose that's fine? I thought it could be problematic sometimes.

whsieh replied:
Oh, right — I meant to use `BOOL` for the value here (although implicit conversion is technically okay in this case). AFAIK, the only case where the implicit conversion causes problems is on Catalyst, if the compiler tries to determine the implicit return type of a C++ lambda that has branches which return both a `bool` and a `BOOL`. A simple example of this is something like:

```
auto result = [&] {
    if (returnBool())
        return false;

    BOOL value = getValue();
    return value;
}();
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0c5e147fdafcdcf5fa9be33fcb69f575367715fc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37333

whsieh commented:
Thanks for the review!

whsieh committed:
[iPadOS] Outlook native app: channels tab does not handle trackpad two-finger pans to scroll
https://bugs.webkit.org/show_bug.cgi?id=283931
rdar://139834901

Reviewed by Abrar Rahman Protyasha.

On iPad, when scrolling via 2-finger swipe on a connected trackpad, `wheel` events are dispatched to
the page via the `WebPage::dispatchWheelEventWithoutScrolling` IPC message. As the name suggests,
this only dispatches `wheel` DOM events to the page, and returns a boolean result back to the UI
process indicating whether or not scrolling should be allowed to commence. UIKit will then only
drive the default scroll animation on the native scroll view under the pointer if the event was not
`handled`.

Whether the event is handled or not depends on the return value of `EventHandler::handleWheelEvent`.
In the case of the Outlook native app's Channels tab, which is comprised of a large full-viewport
subscrollable overflow container, it returns `true` only because the body and root elements have
`overscroll-behavior: none;`, causing us to exit early in `handleWheelEventInternal` with `true`:

```
HandleUserInputEventResult EventHandler::handleWheelEventInternal(…)
{
…
    auto adjustedWheelEvent = event;
    auto filteredDelta = adjustedWheelEvent.delta();
    filteredDelta = view->deltaForPropagation(filteredDelta);
    if (view->shouldBlockScrollPropagation(filteredDelta))
        return true;                                            // <---
```

...and subsequently causes us to block scrolling on the subscrollable container. In this same
codepath on macOS, we also return `true` here, but the key difference is that treating the wheel
event as `handled := true` here doesn't simply block scrolling — that only happens if we've blocked
scrolling due to `preventDefault` being called.

To fix this impedance mismatch between the return value of `handleWheelEvent()` and how it's
interpreted within `WebPage::dispatchWheelEventWithoutScrolling`, we plumb the `EventHandling`
result alongside `HandleUserInputEventResult` up from WebCore to WebKit as a `std::pair`, and
consult the event handling result (i.e. whether or not `EventHandling::DefaultPrevented` is set) in
addition to checking whether the event was handled.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleWheelEvent):

Make this return a `std::pair<HandleUserInputEventResult, OptionSet<EventHandling>>`, and plumb it
through to the client layer (see `WebPage` changes below) so that we can check if the default was
prevented (not just if the event was handled) in `dispatchWheelEventWithoutScrolling`.

(WebCore::EventHandler::passWheelEventToWidget):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::wheelEvent):
* Source/WebCore/page/mac/EventHandlerMac.mm:
(WebCore::EventHandler::wheelEvent):
(WebCore::EventHandler::passWheelEventToWidget):
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::shouldBlockScrollPropagation const):

Drive-by fix: rewrite this method to be more readable:

-   Cache the results of calling `(horizontal|vertical)OverscrollBehaviorPreventsPropagation` in
    local variables instead of calling the respective methods multiple times.

-   Split the return value into several early returns that achieve the same result, but (more
    clearly) represent the scenarios in which this is true: either both axes stop propagation, or
    only one axis stops propagation but the biased delta amount for the opposite axis is 0.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView scrollView:handleScrollUpdate:completion:]):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::dispatchWheelEvent):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::handleWheelEvent):
(WebKit::WebPage::wheelEvent):
(WebKit::WebPage::dispatchWheelEventWithoutScrolling):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Rename this variable to reflect that this result flag represents whether or not scrolling should be
prevented, not just whether the event was handled.

* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
(-[WKWebView synchronouslyHandleScrollEventWithPhase:location:delta:]):
(TEST(WKScrollViewTests, AsynchronousWheelEventHandling)):
(TEST(WKScrollViewTests, OverscrollBehaviorShouldNotPreventScrolling)):

Add a new API test to exercise this change, by verifying that scrolling is not prevented when
dispatching wheel events over an overflow scrollable region inside of a root and body element with
`overscroll-behavior: contain;`. Also verify that scrolling is (correctly) prevented if the page
explicitly calls `preventDefault()` on the wheel event.

Canonical link: https://commits.webkit.org/287293@main
